[FLINK-10977][table] Add streaming non-window FlatAggregate to Table API

hequn8128 opened the pull request:
## What is the purpose of the change

This pull request supports (streaming, nonwindow)flatAggregate on Table API. Note: `emitValueWithRetract`(descriped in [FLIP-29](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=97552739)) will be supported in a later PR which is used to improve the performance.

## Brief change log

- Add user defined table agg function interface
- Add flatAggregate api on table
- Resolve expression and build LogicalTableAggregate
- Add Plan support for table aggregate
- Add code gen support for table aggregate
- Add runtime support for table aggregate
- Add registration for table aggregate


## Verifying this change


This change added tests and can be verified as follows:

 - Added plan test(TableAggregateTest) to verify plan
 - Add IT case(TableAggregateITCase) to verify code gen and runtime logic
 - Add Harness test(TableAggregateHarnessTest) to test runtime logic
 - Add TableAggregateStringExpressionTest to test java.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

sunjincheng121 commented:
Thanks for your great work! 
@flinkbot approve-until architecture
@flinkbot attention @aljoscha @twalthr

sunjincheng121 commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/operations/TableAggFunctionCallFactory.java:
```diff
@@ -0,0 +1,146 @@
+ * Utility class for creating a valid {@link TableAggFunctionCall} operation.
+ */
+@Internal
+public class TableAggFunctionCallFactory {
```
Why we need the factory class, can we add the `ExpressionBridge` as a parameter of `TableAggFunctionCallVisitor#createXX`  methods, or defined `bridge` as a member of `TableAggFunctionCallVisitor`?

hequn8128 replied:
You are right. It's ok to remove the factory class and use visitor directly. I add this just to keep it consistent with the `CalculatedTableFactory` which is used to process udtf.

hequn8128 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/functions/utils/TableAggSqlFunction.scala:
```diff
@@ -0,0 +1,114 @@
+}
+
+
+object TableAggSqlFunction {
```
Yes, it's good to share logic with `AggSqlFunction`. The problem here is `TableAggSqlFunction` contains an alias property. However, I can put alias in `AggSqlFunction` and unify them if you think it makes sense.

hequn8128 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/expressions/aggregations.scala:
```diff
@@ -412,8 +428,28 @@ case class AggFunctionCall(
+  * Expression for calling a user-defined table aggregate function.
+  */
+case class TableAggFunctionCall(
+  aggregateFunction: TableAggregateFunction[_, _],
```
Yes, once we unify `AggSqlFunction` we don't need this `TableAggFunctionCall`.

hequn8128 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/AggregateFunctionDefinition.java:
```diff
@@ -31,13 +31,13 @@
 	public AggregateFunctionDefinition(
 			String name,
-			AggregateFunction<?, ?> aggregateFunction,
+			UserDefinedAggregateFunction<?, ?> aggregateFunction,
```
I'm not sure if it's right to add a new type of FunctionDefinition. TableAggregate can somehow also be treated as an Aggregate function. Maybe we can add this type later if we find it's needed. What do you think?

sunjincheng121 replied:
Yes, make sense to me if we can share more logic between Agg and Table Agg. and we do not need to add a new type.

hequn8128 commented:
@sunjincheng121 Thanks a lot for the review. I think you make a very good point. It's better if we can reuse the original code as much as possible. I will try to address your comments asap.  

Best, Hequn

sunjincheng121 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/codegen/BaseAggregationCodeGenerator.scala:
```diff
@@ -0,0 +1,776 @@
+  *                               method
+  * @param accConfig              Data view specification for accumulators
+  */
+class BaseAggregationCodeGenerator(
```
I like the idea of adding a base class of Aggregate code generator, But I think we can use the  `AggregationCodeGenerator` and only add `generateTableAggregations` is enough for table aggregate code gen. Because if we want to add a  base class of Aggregate code generator, we should consider how to deal with the` aggregates` vs `aggregate`, `distinctAccMapping` which table Aggregate never using.
So, I suggest if we really need add a base class of  Aggregate code generator, we can open a new PR.  What do you think?

hequn8128 commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/operations/AggregateOperationFactory.java:
```diff
@@ -465,4 +535,98 @@ protected Void defaultMethod(Expression expression) {
+			.filter(p -> p.getFunctionDefinition() instanceof AggregateFunctionDefinition)
+			.map(p -> (AggregateFunctionDefinition) p.getFunctionDefinition())
+			.filter(p -> p.getAggregateFunction() instanceof TableAggregateFunction)
+			.collect(Collectors.toList()).size() == 1;
```
Thank you for your suggestions. I will keep the original code. I also make it more simply with `Stream.of` and `anyMatch`.  :-)

hequn8128 commented:
@sunjincheng121 Hi, Thank you very much for your review and great suggestions. I think the pr improved a lot with your help. I have addressed your comments and updated the PR. Would be great if you can take another look. 

Best, Hequn

sunjincheng121 commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/StreamTableEnvironment.java:
```diff
@@ -74,6 +75,17 @@
+	 * catalog. Registered functions can only be referenced in Table API.
+	 *
+	 * @param name The name under which the function is registered.
+	 * @param aggregateFunction The TableAggregateFunction to register.
```
`aggregateFunction` -> `tableAggregateFunction`

sunjincheng121 commented on docs/dev/table/tableApi.md:
```diff
@@ -1892,6 +1892,59 @@ Table table = input
+        <span class="label label-info">Result Updating</span>
+      </td>
+      <td>
+        <p>Similar to a <b>GroupBy Aggregation</b>. Groups the rows on the grouping keys with a following running table aggregation operator to aggregate rows group-wise. The difference from an AggregateFunction is that TableAggregateFunction may return 0 or more records for a group. You have to close the "flatAggregate" with a select statement. And the select statement does not support aggregate ...
```
`with a following` -> ` with the following`

sunjincheng121 commented on docs/dev/table/tableApi.md:
```diff
@@ -1892,6 +1892,59 @@ Table table = input
+    .flatAggregate("myTableAggFunc(a) as (x, y)")
+    .select("a, x, y");
+{% endhighlight %}
+        <p><b>Note:</b> For streaming queries the required state to compute the query result might grow infinitely depending on the type of aggregation and the number of distinct grouping keys. Please provide a query configuration with valid retention interval to prevent excessive state size. See <a href="streaming/query_configuration.html">Query Configuration</a> for details.</p>
```
`with valid` -> `with a valid`?

sunjincheng121 replied:
`For streaming queries ` -> `For streaming queries,`

hequn8128 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -227,6 +226,79 @@ object AggregateUtil {
+    * Create an [[org.apache.flink.streaming.api.functions.ProcessFunction]] for group (without
+    * window) aggregate to evaluate final table aggregate value.
+    *
+    * @param generator       code generator instance
```
good catch!

hequn8128 commented:
@sunjincheng121 Thank you very much for your continues review and suggestions. I have addressed all your comments and updated the PR. :-)

Best, Hequn

sunjincheng121 commented:
Thanks for your update!
Then changes look good to me now!

And one more thing is in current PR without the implementation of incremental mode, can you create a JIRA for incremental mode?

Best,Jincheng

hequn8128 commented:
@sunjincheng121  Thank you for the reminder. I have created another jira([FLINK-12401](https://issues.apache.org/jira/browse/FLINK-12401)) for the incremental mode.

Best, Hequn

sunjincheng121 commented:
Thanks for create the JIRA @hequn8128!
+1 to merged.
@flinkbot approve all
